### PR TITLE
Big refactoring (v. 2)

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | -

Before merge:
- [x] Merge [PR #121](https://github.com/yiisoft/html/pull/121) in Yii HTML
- [x] Release Yii HTML 2.4.0
- [x] Add `yiisoft/html:^2.4.0` to requirements of demo (in adopt PR)
- [x] Add `yiisoft/html:^2.4.0` to requirements in this PR

Adopt PRs:
- https://github.com/yiisoft/demo/pull/454
- https://github.com/yiisoft/yii-bulma/pull/64
- https://github.com/yiisoft/yii-bootstrap5/pull/88

### CODE REVIEW

**Review by rustamwin — COMMENTED** [review]
Looks good. What about move reusable methods like "ariaLabel", "autofocus", "tabIndex" to traits?

**Review by rustamwin — COMMENTED** [review]
Looks good. What about moving reusable methods like "ariaLabel", "autofocus", "tabIndex" to traits?

File: src/Field/Base/PartsField.php [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+                sprintf(
+                    'Token name "%s" is built-in.',
+                    $token,
+                )
+            );
+        }
+
+        if ($token === '') {
+            throw new InvalidArgumentException('Token must be non-empty string.');
+        }
```
[the remark this answers is not in the document]
Comment (vjik) [reply src/Field/Base/PartsField.php]: Done

**Comment by vjik** [comment]
> Looks good. What about moving reusable methods like "ariaLabel", "autofocus", "tabIndex" to traits?

It is can help write code, but this approach has its drawbacks:
- in some fields same methods requires different parameters (for example, in `Date` field method `max` requires string, but in `Number` field requires string, integer or float);
- harder explore code (on go to method in IDE better to be in the class than in the trait);
- more complex code.

My opinion in this case — don't use traits and allow a little duplication.

File: src/Field/Email.php [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+    /**
+     * Allow to specify more than one value.
+     *
+     * @param bool $value Whether the user is to be allowed to specify more than one value.
+     *
+     * @link https://html.spec.whatwg.org/multipage/input.html#attr-input-multiple
+     */
+    public function multiple(bool $value = true): self
```
Comment (samdark) [inline src/Field/Email.php:69]: For boolean fields let's ensure there's methods like this accepting value and no methods like `withoutX()`. I've seen these in other classes (do search "on without" in IDE please).

**Review by samdark — CHANGES_REQUESTED** [review]

File: docs/creating-fields.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
+# Creating and Using Custom Fields
+
+Using the fields included in the package, you may also create your own fields.
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/creating-fields.md:3]: Replaceв with "You may create your own fields based on the **abstract** fields included in the package."

File: docs/creating-fields.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+## Creating Field Class
+
+Field class must extend `Yiisoft\Form\Field\Base\BaseField` or one of his heirs and implement necessary abstract 
+methods.
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/creating-fields.md:8]: done

File: docs/creating-fields.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+Using the fields included in the package, you may also create your own fields.
+
+## Creating Field Class
+
+Field class must extend `Yiisoft\Form\Field\Base\BaseField` or one of his heirs and implement necessary abstract 
+methods. 
+
+### Base Classes
+
+- `BaseField` — base class that contain common functionality of fields.
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/creating-fields.md:12]: done

File: docs/creating-fields.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+- `PartsField` — class extends `BaseField`, add templating functionality and parts of field (label, hint, error).  
+- `InputField` — class extends `PartsField`, add form model support (in most cases, this class is used).
+- `ButtonField` — specific class for button fields, extends `PartsField`.
+- `DateTimeInputField` — specific class for creating form controls with date and/or time input, extends `InputField`.
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/creating-fields.md:16]: done

File: docs/creating-fields.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+  PartsField  -->  BaseField 
+```
+
+### Feature Traits
+
+You can use feature traits for your field class.
+
+#### `FormAttributeTrait` 
+
+Add methods for using form model.
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/creating-fields.md:34]: done

File: docs/creating-fields.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+#### `FieldContentTrait`
+
+Add methods for set and generate custom content of fields.
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/creating-fields.md:38]: done

File: docs/creating-fields.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+#### `VaidationClassTrait` 
+
+Add methods for set valid and invalid CSS classes. 
+
+For applying common fields configuration when field creating through a field factory class must implement 
+`ValidationClassInterface`.
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/creating-fields.md:45]: done

File: docs/creating-fields.md [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+#### `PlaceholderTrait`
+
+Add methods for using placeholder. Available for heirs of `InputField` only. 
+
+For applying common fields configuration when field creating through a field factory class must implement
+`PlaceholderInterface`.
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/creating-fields.md:52]: done

File: docs/creating-fields.md [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+#### `EnrichmentFromRulesTrait`
+
+Add method for set option of enrichment field from form model rules.
+
+For applying common fields configuration when field creating through a field factory class must implement
+`EnrichmentFromRulesInterface`.
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/creating-fields.md:59]: done

File: docs/creating-fields.md [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+## Field Configuration
+
+Set base configuration of field in parameter "fieldConfigs" of field factory. For example:
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/creating-fields.md:63]: done

File: docs/creating-fields.md [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+            $html .= '<span class="input-group-text">' . Html::encode($this->suffix) . '</span>' . "\n";
+            $html .= '</div>';
+        }
+
+        return $html;
+    }
+}
+```
+
+Field using:
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/creating-fields.md:119]: done

File: docs/field-parts/error.md [hunk 14]
```diff
@@ -0,0 +1,42 @@
+# Error Field Part
+
+Represent a field validation error (if there are several errors, the first one is used). If field is no validation
+error, field part will be hidden.
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/field-parts/error.md:4]: done

File: docs/fields-configuration.md [hunk 15]
```diff
@@ -0,0 +1,8 @@
+# Fields configuration
+
+Configuration levels:
+
+- widget factory configuration;
+- common configuration in field factory;
+- individual configuration in field factory;
+- configuration at the place of use.
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields-configuration.md:8]: Yes. I plan to add docs in another PR.

File: docs/fields/button.md [hunk 16]
```diff
@@ -0,0 +1,22 @@
+# Button Field
+
+Represents `<button>` element of type "button" are rendered as button without default behavior. Documentation:
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields/button.md:3]: done

File: docs/fields/date-time-local.md [hunk 17]
```diff
@@ -0,0 +1,56 @@
+# DateTimeLocal Field
+
+Represents `<input>` element of type "datetime-local" are let the user easily enter both a date and a time, including
+the year, month, and day as well as the time in hours and minutes. Documentation:
+
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields/date-time-local.md:5]: done

File: docs/fields/date-time.md [hunk 18]
```diff
@@ -0,0 +1,53 @@
+# DateTime Field
+
+Represents `<input>` element of type "datetime" are let the user enter a date and time (hour, minute, second, and
+fraction of a second) as well as a timezone. Documentation:
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields/date-time.md:4]: done

File: docs/fields/date.md [hunk 19]
```diff
@@ -0,0 +1,56 @@
+# Date Field
+
+Represents `<input>` element of type "date" are let the user enter a date, either with a textbox that validates
+the input or a special date picker interface. Documentation:
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields/date.md:4]: done

File: docs/fields/email.md [hunk 20]
```diff
@@ -0,0 +1,47 @@
+# Email Field
+
+Represents `<input>` element of type "email" are let the user enter and edit an e-mail address, or, if the "multiple"
+attribute is specified, a list of e-mail addresses. Documentation:
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields/email.md:4]: done

File: docs/fields/fieldset.md [hunk 21]
```diff
@@ -0,0 +1,43 @@
+# Fieldset Field
+
+Represents `<fieldset>` element are use to group several controls. Documentation:
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields/fieldset.md:3]: done

File: docs/fields/file.md [hunk 22]
```diff
@@ -0,0 +1,48 @@
+# File Field
+
+Represents `<input>` element of type "file" are ley the user choose one or more files from their device storage.
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields/file.md:3]: done

File: docs/fields/hidden.md [hunk 23]
```diff
@@ -0,0 +1,36 @@
+# Hidden Field
+
+Represents `<input>` element of type "hidden" are let web developers include data that cannot be seen or modified by
+users when a form is submitted. Documentation:
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields/hidden.md:4]: done

File: docs/fields/image.md [hunk 24]
```diff
@@ -0,0 +1,24 @@
+# Image Field
+
+Represents `<input>` element of type "image" are used to create graphical submit buttons. Documentation:
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields/image.md:3]: done

File: docs/fields/number.md [hunk 25]
```diff
@@ -0,0 +1,47 @@
+# Number Field
+
+Represents `<input>` element of type "number" are let the user enter a number. Documentation:
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields/number.md:3]: done

File: docs/fields/password.md [hunk 26]
```diff
@@ -0,0 +1,46 @@
+# Password Field
+
+Represents `<input>` element of type "password" are let the user securely enter a password. Documentation:
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields/password.md:3]: done

File: docs/fields/range.md [hunk 27]
```diff
@@ -0,0 +1,49 @@
+# Range Field
+
+Represents `<input>` element of type "range" are let the user specify a numeric value which must be no less than a given
+value, and no more than another given value. Documentation:
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields/range.md:4]: done

File: docs/fields/reset-button.md [hunk 28]
```diff
@@ -0,0 +1,23 @@
+# Reset Button Field
+
+Represents `<button>` element of type "reset" are rendered as button for resets all the controls to their initial
+values. Documentation:
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields/reset-button.md:4]: done

File: docs/fields/submit-button.md [hunk 29]
```diff
@@ -0,0 +1,22 @@
+# Submit Button Field
+
+Represents `<button>` element of type "submit" are rendered as button for submitting a form. Documentation:
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields/submit-button.md:3]: done

File: docs/fields/telephone.md [hunk 30]
```diff
@@ -0,0 +1,46 @@
+# Telephone Field
+
+Represents `<input>` element of type "tel" are let the user enter and edit a telephone number. Documentation:
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields/telephone.md:3]: done

File: docs/fields/url.md [hunk 31]
```diff
@@ -0,0 +1,44 @@
+# Url Field
+
+Represents `<input>` element of type "url" are let the user enter and edit a URL. Documentation:
```
[the remark this answers is not in the document]
Comment (vjik) [reply docs/fields/url.md:3]: done

File: src/Field.php [hunk 32]
```diff
@@ -0,0 +1,241 @@
...
+
+    /**
+     * @psalm-var array<string,FieldFactory>
+     */
+    private static array $factories = [];
+
+    /**
+     * @psalm-param array<string,array> $configs
+     */
+    public static function initialize(array $configs = [], string $defaultConfigName = 'default'): void
```
[the remark this answers is not in the document]
Comment (vjik) [reply src/Field.php:57]: Yes. For example, user can change default config name for own application to "bulma" or "bootstrap5".

File: src/Field/Base/BaseField.php [hunk 33]
```diff
@@ -0,0 +1,184 @@
...
+        $this->beforeRender();
+
+        $content = $this->generateBeginContent();
+
+        $html = $this->renderOpenContainerAndContent($content);
+        if ($html === '') {
+            return '';
+        }
+
+        return $html;
```
[the remark this answers is not in the document]
Comment (vjik) [reply src/Field/Base/BaseField.php:108]: done

File: src/Field/Base/EnrichmentFromRules/EnrichmentFromRulesInterface.php [hunk 34]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Form\Field\Base\EnrichmentFromRules;
+
+interface EnrichmentFromRulesInterface
```
[the remark this answers is not in the document]
Comment (vjik) [reply src/Field/Base/EnrichmentFromRules/EnrichmentFromRulesInterface.php:7]: Added phpdoc.

File: src/Field/Base/FieldContentTrait.php [hunk 35]
```diff
@@ -0,0 +1,95 @@
...
+        $new->content = array_values($content);
+        return $new;
+    }
+
+    /**
+     * @param string|Stringable ...$content Field content.
+     *
+     * @return static
+     */
+    final public function addContent(string|Stringable ...$content): self
```
[the remark this answers is not in the document]
Comment (vjik) [reply src/Field/Base/FieldContentTrait.php:71]: Yes. This logic from Yii HTML.

Attributes always [merge by default](https://github.com/yiisoft/html/blob/b0c0729e1f8516cafcafec2af103405fab39eb81/src/Tag/Base/Tag.php#L25), another properties replace by default,[ including content](https://github.com/yiisoft/html/blob/b0c0729e1f8516cafcafec2af103405fab39eb81/src/Tag/Base/TagContentTrait.php#L59).

But in Yii HTML logic will be changed in future version: https://github.com/yiisoft/html/issues/122.

I think in forms need rename all `attributes()` to `addAttributes()` and `replaceAttributes()` to `attributes()`. It's OK?

File: src/Field/Email.php (hunk 2 above)
Comment (vjik) [reply src/Field/Email.php:69]: I don't found methods `without*()`

File: src/Field.php (hunk 32 above)
[the remark this answers is not in the document]
Comment (samdark) [reply src/Field.php:57]: Should be mentioned in docs, I guess.

File: src/Field/Base/FieldContentTrait.php (hunk 35 above)
[the remark this answers is not in the document]
Comment (samdark) [reply src/Field/Base/FieldContentTrait.php:71]: Yes. Fine.

[the remark this answers is not in the document]
Comment (vjik) [reply src/Field/Base/FieldContentTrait.php]: Did it. Also renamed "class" methods by similar way. Added issues to Yii HTML:
- https://github.com/yiisoft/html/issues/123
- https://github.com/yiisoft/html/issues/124
- https://github.com/yiisoft/html/issues/122
- https://github.com/yiisoft/html/issues/125

File: src/Field.php (hunk 32 above)
[the remark this answers is not in the document]
Comment (vjik) [reply src/Field.php:86]: Added phpdoc

**Review by samdark — APPROVED** [review]

**Pull request merged**